src/com/amazon/ion/impl/ResizingPipedInputStream.java

```diff
@@ -0,0 +1,557 @@
+     * each time it grows.
+     */
+    private final int initialBufferSize;
+
+    /**
+     * The maximum size of the buffer. If the user attempts to buffer more bytes than this, an exception will be raised.
+     */
+    private final int maximumBufferSize;
+
+    /**
+     * Whether to use a boundary to limit the number of available bytes. This can be used to buffer
+     * arbitrarily-sized chunks of bytes without making them available for consumption. When true,
+     * the boundary must be manually extended (see {@link #extendBoundary(int)} to make these bytes
+     * available. When false, all buffered bytes will be available to read.
+     */
+    private final boolean useBoundary;
+
+    /**
+     * The NotificationConsumer currently registered.
+     */
+    private NotificationConsumer notificationConsumer = NO_OP_NOTIFICATION_CONSUMER;
+
+    /**
+     * The raw buffer.
+     */
+    private byte[] buffer;
+
+    /**
+     * View to the raw buffer.
+     */
+    private ByteBuffer byteBuffer;
+
+    /**
+     * @see #capacity()
+     */
+    private int capacity;
+
+    /**
+     * The index of the next byte in the buffer that is available to be read. Always less than or equal to `writeIndex`.
+     */
+    private int readIndex = 0;
+
+    /**
+     * The index at which the next byte received will be written. Always greater than or equal to `readIndex`.
+     */
+    private int writeIndex = 0;
+
+    /**
+     * @see #available()
+     */
+    private int available = 0;
+
+    /**
+     * @see #size()
+     */
+    private int size = 0;
+
+    /**
+     * @see #getBoundary()
+     */
+    private int boundary = 0;
+
+    /**
+     * Constructor.
+     * @param initialBufferSize the initial size of the buffer. When full, the buffer will grow by this
+     *                          many bytes. The buffer always stores bytes contiguously, so growth requires
+     *                          allocation of a new buffer capable of holding the new capacity and copying of the
+     *                          existing bytes into the new buffer. As such, a size should be chosen carefully
+     *                          such that growth is expected to occur rarely, if ever.
+     */
+    public ResizingPipedInputStream(final int initialBufferSize) {
+        this(initialBufferSize, Integer.MAX_VALUE, false);
+    }
+
+    /**
+     * Constructor.
+     * @param initialBufferSize the initial size of the buffer. When full, the buffer will grow by this
+     *                          many bytes. The buffer always stores bytes contiguously, so growth requires
+     *                          allocation of a new buffer capable of holding the new capacity and copying of the
+     *                          existing bytes into the new buffer. As such, a size should be chosen carefully
+     *                          such that growth is expected to occur rarely, if ever.
```

Comment:
The javadoc style doesn't really matter to me either way. The current style is what IntelliJ does for me when I go to a new line within a Javadoc parameter description, which is why it's here. I like the way you've proposed slightly more. Since there are many other comments on this PR I'm going to focus on those, and we can follow up with a change to the Javadoc style in a future PR.

Reply:
https://github.com/amzn/ion-java/issues/376